File: Mirror_Method/func_FilterDfactor.py
import numpy as np
import matplotlib.pyplot as plt

# 以前のコードと同じ DFactorFilter クラスをここに配置する
class DFactorFilter:
    """
    資料の9.3節で解説されている二相入力・二相出力のD因子フィルタを実装したクラス。
    """
    def __init__(self, a0, b0, b1, omega0, fs):
        self.a0 = a0
        self.b0 = b0
        self.b1 = b1
        self.omega0 = omega0
        self.Ts = 1.0 / fs
        self.x_p = 0 + 0j
        
    def process(self, u):
        u_p = u[0] + 1j * u[1]
        x_p_new = (self.x_p + self.Ts * u_p) / (1 + (self.a0 + 1j * self.omega0) * self.Ts)
        y_p = (self.b0 - self.a0 * self.b1) * x_p_new + self.b1 * u_p
        self.x_p = x_p_new
        return np.array([y_p.real, y_p.imag])

# --- メイン実行部 ---
if __name__ == '__main__':
    # シミュレーション設定
    FS = 12000  # サンプリング周波数: 12kHz
    T = 0.05    # シミュレーション時間 (s)
    t = np.arange(0, T, 1.0/FS)

    # フィルタのパラメータ設定
    # F(s) = 100/(s+100) -> a0=100, b0=100, b1=0
    # このフィルタは帯域幅が約100 rad/s (約16Hz)のローパスフィルタ
    A0 = 100.0
    B0 = 100.0
    B1 = 0.0
    
    # シフト周波数: -1kHz -> -2*pi*1000 rad/s
    # (通過帯域の中心が 1kHz になるように負値を設定)
    OMEGA0_SHIFT = -2 * np.pi * 1000.0

    # D因子フィルタのインスタンスを作成
    d_filter = DFactorFilter(a0=A0, b0=B0, b1=B1, omega0=OMEGA0_SHIFT, fs=FS)

    # テスト入力信号の生成 (1kHz信号 + 3kHzノイズ)
    freq_target = 1000  # Hz (フィルタで抽出したい周波数)
    w_target = 2 * np.pi * freq_target
    signal = np.array([np.cos(w_target * t), np.sin(w_target * t)])

    freq_noise = 3000   # Hz (除去したいノイズ周波数)
    w_noise = 2 * np.pi * freq_noise
    noise = np.array([np.cos(w_noise * t), np.sin(w_noise * t)]) * 0.5
    
    input_signal = (signal + noise).T

    # フィルタ処理の実行
    output_signal = np.zeros_like(input_signal)
    for i in range(len(t)):
        output_signal[i, :] = d_filter.process(input_signal[i, :])

    # 結果のプロット
    plt.figure(figsize=(12, 6))
    plt.plot(t, input_signal[:, 0], 'c-', label='Input u1 (1kHz + 3kHz Noise)', alpha=0.7)
    plt.plot(t, output_signal[:, 0], 'b-', label='Output y1 (1kHz Filtered)')
    plt.title("D-Factor Filter Response (Fs=12kHz, F_shift=1kHz)")
    plt.xlabel("Time (s)")
    plt.ylabel("Amplitude")
    plt.xlim(0, 0.01) # 波形が見やすいように時間範囲を拡大
    plt.grid(True)
    plt.legend()
    plt.show()

File: Mirror_Method/func_FilterDfactor.m
function y = DFactorFilter_matlab(u)
% DFactorFilter_matlab
% 資料9.3節に基づく二相入力・二相出力D因子フィルタのMATLAB実装
%
%   Input:
%       u: 2x1の入力信号ベクトル [u1; u2]
%   Output:
%       y: 2x1の出力信号ベクトル [y1; y2]
%
%   内部で状態を保持するため、連続して呼び出すことでフィルタとして機能します。

% --- 設定 (関数内定義) ---
persistent x_p; % フィルタの内部状態を保持する永続変数

% パラメータ (固定値)
FS = 12000; % サンプリング周波数: 12kHz
A0 = 100.0; % F(s)の係数 a0
B0 = 100.0; % F(s)の係数 b0
B1 = 0.0;   % F(s)の係数 b1

% シフト周波数: 1kHz -> 2*pi*1000 rad/s
% 通過帯域の中心が1kHzになるように負値を設定
OMEGA0_SHIFT = -2 * pi * 1000.0;

% --- 永続変数の初期化 (初回実行時のみ) ---
if isempty(x_p)
    x_p = 0 + 0j;
end

% --- フィルタ処理 ---
Ts = 1.0 / FS;

% 1. 分解: 2相実数信号を複素数に変換
u_p = u(1) + 1j * u(2);

% 2. 複素フィルタリング (伝達関数と状態方程式に基づく離散化)
% s*x = -(a0 + j*omega0)*x + u_p
% これを離散化（後退差分）: (x_new - x_old)/Ts = -(a0+j*w0)*x_new + u_p
% (1 + (a0+j*w0)*Ts)*x_new = x_old + Ts*u_p
x_p_new = (x_p + Ts * u_p) / (1 + (A0 + 1j * OMEGA0_SHIFT) * Ts);

% 出力方程式 Y = (b0 - a0*b1)*x + b1*U
y_p = (B0 - A0 * B1) * x_p_new + B1 * u_p;

% 状態の更新
x_p = x_p_new;

% 3. 再構成: 複素数を出力2相実数信号に変換
y = [real(y_p); imag(y_p)];

end